File: ios/RoutineChart/RoutineChart/Core/Domain/UseCases/CreateRoutineUseCase.swift
//
//  CreateRoutineUseCase.swift
//  RoutineChart
//
//  Use case for creating a routine with steps
//

import Foundation
import OSLog

final class CreateRoutineUseCase {
    private let routineRepository: RoutineRepository
    private let stepRepository: RoutineStepRepository
    
    init(
        routineRepository: RoutineRepository,
        stepRepository: RoutineStepRepository
    ) {
        self.routineRepository = routineRepository
        self.stepRepository = stepRepository
    }
    
    func execute(
        userId: String,
        familyId: String? = nil,
        title: String,
        iconName: String? = nil,
        steps: [(label: String?, iconName: String?)]
    ) async throws -> Routine {
        // Create routine
        let routine = Routine(
            userId: userId,
            familyId: familyId,
            title: title,
            iconName: iconName
        )
        
        try await routineRepository.create(routine)
        
        // Create steps
        for (index, stepData) in steps.enumerated() {
            let step = RoutineStep(
                routineId: routine.id,
                orderIndex: index,
                label: stepData.label,
                iconName: stepData.iconName
            )
            try await stepRepository.create(step)
        }
        
        AppLogger.domain.info("Created routine '\(title)' with \(steps.count) steps")
        
        return routine
    }
}

File: android/app/src/main/java/com/HammersTech/RoutineChart/core/domain/usecases/CreateRoutineUseCase.kt
package com.HammersTech.RoutineChart.core.domain.usecases

import com.HammersTech.RoutineChart.core.domain.models.CompletionRule
import com.HammersTech.RoutineChart.core.domain.models.Routine
import com.HammersTech.RoutineChart.core.domain.models.RoutineStep
import com.HammersTech.RoutineChart.core.domain.repositories.RoutineRepository
import com.HammersTech.RoutineChart.core.domain.repositories.RoutineStepRepository
import com.HammersTech.RoutineChart.core.utils.AppLogger
import java.time.Instant
import java.util.UUID
import javax.inject.Inject

/**
 * Use case for creating a routine with steps
 */
class CreateRoutineUseCase @Inject constructor(
    private val routineRepository: RoutineRepository,
    private val routineStepRepository: RoutineStepRepository
) {
    data class StepInput(
        val label: String?,
        val iconName: String?,
        val audioCueUrl: String? = null
    )

    suspend operator fun invoke(
        userId: String,
        familyId: String? = null,
        title: String,
        iconName: String?,
        steps: List<StepInput>
    ): Routine {
        val routineId = UUID.randomUUID().toString()
        val now = Instant.now()

        // Create routine
        val routine = Routine(
            id = routineId,
            userId = userId,
            familyId = familyId,
            title = title,
            iconName = iconName,
            version = Routine.DEFAULT_VERSION,
            completionRule = CompletionRule.ALL_STEPS_REQUIRED,
            createdAt = now,
            updatedAt = now,
            deletedAt = null
        )

        routineRepository.create(routine)

        // Create steps (no familyId needed)
        steps.forEachIndexed { index, stepInput ->
            val step = RoutineStep(
                id = UUID.randomUUID().toString(),
                routineId = routineId,
                orderIndex = index,
                label = stepInput.label,
                iconName = stepInput.iconName,
                audioCueUrl = stepInput.audioCueUrl,
                createdAt = now,
                deletedAt = null
            )
            routineStepRepository.create(step)
        }

        AppLogger.UseCase.info("Created routine: $title with ${steps.size} steps")
        return routine
    }
}

